src/types/redis_cms.cc

```diff
@@ -0,0 +1,252 @@
+  }
+  if (!s.ok()) {
+    return s;
+  }
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisCountMinSketch);
+  batch->PutLogData(log_data.Encode());
+
+  CMSketch cms(metadata.width, metadata.depth, metadata.counter, metadata.array);
+
+  for (auto &element : elements) {
+    if (element.second > 0 && metadata.counter > std::numeric_limits<uint64_t>::max() - element.second) {
+      return rocksdb::Status::InvalidArgument("Overflow error: IncrBy would result in counter overflow");
+    }
+    cms.IncrBy(element.first, element.second);
+    metadata.counter += element.second;
+  }
+
+  metadata.array = std::move(cms.GetArray());
+
+  std::string bytes;
+  metadata.Encode(&bytes);
+  batch->Put(metadata_cf_handle_, ns_key, bytes);
+
+  return storage_->Write(ctx, storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+
+rocksdb::Status CMS::Info(engine::Context &ctx, const Slice &user_key, CMSketch::CMSInfo *ret) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  CountMinSketchMetadata metadata{};
+  rocksdb::Status s = GetMetadata(ctx, ns_key, &metadata);
+
+  if (!s.ok()) {
+    return s;
+  }
+
+  ret->width = metadata.width;
+  ret->depth = metadata.depth;
+  ret->count = metadata.counter;
+
+  return rocksdb::Status::OK();
+};
+
+rocksdb::Status CMS::InitByDim(engine::Context &ctx, const Slice &user_key, uint32_t width, uint32_t depth) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  CountMinSketchMetadata metadata{};
+
+  rocksdb::Status s = GetMetadata(ctx, ns_key, &metadata);
+
+  if (s.ok()) {
+    return rocksdb::Status::InvalidArgument("Key already exists.");
+  }
+
+  if (!s.IsNotFound()) {
+    return s;
+  }
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisCountMinSketch);
+  batch->PutLogData(log_data.Encode());
+
+  metadata.width = width;
+  metadata.depth = depth;
+  metadata.counter = 0;
+  metadata.array = std::vector<uint32_t>(width * depth, 0);
+
+  std::string bytes;
+  metadata.Encode(&bytes);
+  batch->Put(metadata_cf_handle_, ns_key, bytes);
+
+  return storage_->Write(ctx, storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+};
+
+rocksdb::Status CMS::InitByProb(engine::Context &ctx, const Slice &user_key, double error, double delta) {
+  if (error <= 0 || error >= 1) {
+    return rocksdb::Status::InvalidArgument("Error must be between 0 and 1 (exclusive).");
+  }
+  if (delta <= 0 || delta >= 1) {
+    return rocksdb::Status::InvalidArgument("Delta must be between 0 and 1 (exclusive).");
+  }
+  CMSketch::CMSketchDimensions dim = CMSketch::CMSDimFromProb(error, delta);
+  auto s = InitByDim(ctx, user_key, dim.width, dim.depth);
+  if (!s.ok()) {
+    return s;
+  }
+
+  size_t memory_used = dim.width * dim.depth * sizeof(uint32_t);
+  const size_t max_memory = 50 * 1024 * 1024;
```

Comment:
Actually I think this should be put in `InitByDim` before allocating the memory or put to storage